File: Leetcode/house_robber.cpp
// https://leetcode.com/problems/house-robber/

class Solution {
public:
    int rob(vector<int>& nums) {
        vector<int> dp(nums.size(), 0);
        dp[0] = nums[0];
        
        if ( nums.size() >= 2 ){
            dp[1] = max(nums[0], nums[1]);

            for ( int i=2; i<nums.size(); i++){
                for( int j=0; j<i-1; j++ ){
                    dp[i] = max(dp[i-1], nums[i]+dp[j]);
                }

            }
        }

        return dp[nums.size()-1];        
    }
};

File: Leetcode/HouseRobber.java
class Solution 
{
    public int rob(int[] a) 
    {
        int n = a.length;
        if(n==1)
            return a[0];
        if(n==2)
            return Math.max(a[0], a[1]);
        int[] dp = new int[n+2];
        for(int i=n-1; i>=0; i--)
            dp[i] = Math.max(dp[i+2] + a[i], dp[i+1]);
        return dp[0];
    }
}
